keps/3258-delayed-admission-check-retries/README.md

```diff
@@ -0,0 +1,566 @@
+
+## Design Details
+
+### API Changes
+
+Add new fields to `AdmissionCheckState` in `workload_types.go`:
+
+```go
+type AdmissionCheckState struct {
+    // Existing fields remain unchanged...
+
+    // RetryAfterSeconds indicates how long to wait at least before
+    // retrying to admit the workload.
+    // When State=Retry, admission check controllers should set this
+    // to implement delays between retry attempts.
+    //
+    // If nil when State=Retry, Kueue will retry immediately.
+    // If set, Kueue will add the workload back to the queue after
+    //   Eviction time + RetryAfterSeconds
+    //
+    // +optional
+    RetryAfterSeconds *metav1.Duration `json:"retryAfterSeconds,omitempty"`
+
+    // RetryCount tracks retry attempts for this admission check.
+    // Kueue automatically increments this when transitioning from
+    // Retry→Pending. Reset to 0 when check succeeds or workload
+    // is admitted.
+    //
+    // +optional
+    RetryCount *int32 `json:"retryCount,omitempty"`
+}
+```
+
+#### Example: Workload Status with Multiple Admission Checks
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: Workload
+metadata:
+  name: ml-training-job
+  namespace: research
+status:
+  conditions:
+    - type: Evicted
+      status: "True"
+      lastTransitionTime: "2024-02-06T10:10:00Z"
+
+  requeueState:
+    requeueAt: "2024-02-07T00:10:00Z"
+
+  admissionChecks:
+    # Check 1: Budget check with daily reset
+    - name: budget-check
+      state: Retry
+      lastTransitionTime: "2024-02-06T10:30:00Z"
+      message: "Daily budget exhausted. Will retry at midnight."
+      retryAfterSeconds: 50400 // in 14h, which is approx midnight
+      retryCount: 1
+
+    # Check 2: GPU availability with exponential backoff
+    - name: gpu-availability
+      state: Retry
+      lastTransitionTime: "2024-02-06T10:30:00Z"
+      message: "No A100 GPUs available. Retrying in 8m (attempt 4)"
+      retryAfterSeconds: 480 # 8 minutes from now
+      retryCount: 4
+
+    # Check 3: License server check - immediate retry
+    - name: license-check
+      state: Retry
+      lastTransitionTime: "2024-02-06T10:30:00Z"
+      message: "License server temporarily unavailable"
+      # No retryAfterSeconds set - will retry immediately
+# In this example, Kueue will wait until 00:00:00Z (the longest delay)
+# before re-evaluating all admission checks
+```
+
+### Controller Changes
+
+1. **Workload Controller** (`pkg/controller/core/workload_controller.go`):
+   - When processing admission checks in `Retry` state, calculate the maximum `retryAfterSeconds` across all checks
+   - Set `.status.requeueState.requeueAt` based on the calculated delay
+   - Increment `retryCount` when transitioning back to `Pending`
+2. The scheduler (`pkg/scheduler/scheduler.go`) will need updates to:
+   - Respect retry delays when processing workloads
+   - Ensure workloads are not scheduled before their retry time has elapsed
```

Comment:
We don't need to detail really. Or it could go maybe other "Implementation details" subsection or "Design details", to make it clear this is low-level stuff.

Also, I'm not sure what is behind this statement "Respect retry delays when processing workloads". I believe we could just rely on the Requeued=True being the guard to scheduling the workload, without changes directly to scheduler.go

Reply:
I've removed this section. It does not provide any value here.